libr/util/rbtree.c

```diff
@@ -382,3 +384,130 @@ R_API void r_rbtree_iter_next(RBIter *it) {
+	RContRBNode *in_tree_node = container_of (_in_tree, RContRBNode, node);
+	return cmp_wrap->cmp (incoming_node->data, in_tree_node->data, cmp_wrap->user);
+}
+
+static int cont_rbtree_search_cmp_wrapper(const void *incoming, const RBNode *in_tree, void *user) {
+	RCRBCmpWrap *cmp_wrap = (RCRBCmpWrap *)user;
+	RBNode *_in_tree;
+	memcpy (&_in_tree, &in_tree, sizeof (RBNode *));
+	RContRBNode *in_tree_node = container_of (_in_tree, RContRBNode, node);
+	return cmp_wrap->cmp (incoming, in_tree_node->data, user);
+}
+
+static int cont_rbtree_free_cmp_wrapper(const void *data, const RBNode *in_tree, void *user) {
+	RCRBCmpWrap *cmp_wrap = (RCRBCmpWrap *)user;
+	void *dat;
+	memcpy (&dat, &data, sizeof (void *));
+	const int ret = cont_rbtree_cmp_wrapper (dat, in_tree, user);
+	if (!ret && cmp_wrap->free) { //this is for deleting
+		RContRBNode *in_tree_node = container_of (in_tree, RContRBNode, node);
+		cmp_wrap->free (in_tree_node->data);
+	}
+	return ret;
+}
+
+R_API bool r_rbtree_cont_insert(RContRBTree *tree, void *data, RContRBCmp cmp, void *user) {
+	r_return_val_if_fail (tree && cmp, false);
+	if (!tree->root) {
+		tree->root = R_NEW0 (RContRBNode);
+		if (tree->root) {
+			tree->root->data = data;
+			//			tree->root->node.red = false;	// not needed since R_NEW0 initializes with false anyway
+			return true;
+		}
+		eprintf ("Allocation failed\n");
+		return false;
+	}
+	RContRBNode *incoming_node = R_NEW0 (RContRBNode);
+	if (!incoming_node) {
+		eprintf ("Allocation failed\n");
+		return false;
+	}
+	incoming_node->data = data;
+	RCRBCmpWrap cmp_wrap = { cmp, NULL, user };
+	RBNode *root_node = &tree->root->node;
+	const bool ret = r_rbtree_aug_insert (&root_node, incoming_node,
+		&incoming_node->node, cont_rbtree_cmp_wrapper, NULL, &cmp_wrap);
+	if (root_node != (&tree->root->node)) {
+		tree->root = container_of (root_node, RContRBNode, node); //cursed augmentation garbage
+	}
+	if (!ret) {
+		eprintf ("Insertion failed\n");
+		free (incoming_node);
+	}
+	return ret;
+}
+
+static void cont_node_free(RBNode *node) {
+	free (container_of (node, RContRBNode, node));
+}
+
+R_API bool r_rbtree_cont_delete(RContRBTree *tree, void *data, RContRBCmp cmp, void *user) {
+	r_return_val_if_fail (tree && cmp && tree->root, false);
+	RCRBCmpWrap cmp_wrap = { cmp, tree->free, user };
+	RContRBNode data_wrap = { { { NULL, NULL }, false }, data };
+	RBNode *root_node = &tree->root->node;
+	const bool ret = r_rbtree_aug_delete (&root_node, &data_wrap, cont_rbtree_free_cmp_wrapper, cont_node_free, NULL, &cmp_wrap);
+	if (root_node != (&tree->root->node)) {	//can this crash?
+		tree->root = container_of (root_node, RContRBNode, node); //cursed augmentation garbage
+	}
+	if (!ret) {
+		eprintf ("Deletion failed\n");
+	}
+	return ret;
+}
+
+R_API void *r_rbtree_cont_find(RContRBTree *tree, void *data, RContRBCmp cmp, void *user) {
+	r_return_val_if_fail (tree && cmp && tree->root, NULL);
+	RCRBCmpWrap cmp_wrap = { cmp, NULL, user };
```

Comment:
is this the only place where this user is used? because i want to understand your needs to add an extra parameter

Reply:
nope, i also need it in generation of the esil-cfg to filter out dead code. I have at least 2 more occassions in mind, where this is going to be needed, one of them is VSA (a "more fancy graphs with esil"-thing), and the other one is in io. However I don't want to blow up this huge pr even more, so just wait for it to happen. in the